File: drivers/swift/Sources/OrbeliasDB/OrbeliasDB.swift
import COrbeliasDB
import Foundation

public enum OrbeliasDBError: Error, CustomStringConvertible, Equatable {
    case abiVersionMismatch(expected: Int32, actual: Int32)
    case cAbi(String)
    case closed
    case invalidUtf8

    public var description: String {
        switch self {
        case let .abiVersionMismatch(expected, actual):
            return "OrbeliasDB ABI version mismatch: expected \(expected), got \(actual)"
        case let .cAbi(message):
            return message
        case .closed:
            return "OrbeliasDB handle is closed"
        case .invalidUtf8:
            return "OrbeliasDB returned invalid UTF-8"
        }
    }
}

public struct OrbeliasId: Equatable, Sendable {
    public let parent: UInt64
    public let epoch: UInt32
    public let seq: UInt32
    public let tWrite: Double

    public init(parent: UInt64, epoch: UInt32, seq: UInt32, tWrite: Double) {
        self.parent = parent
        self.epoch = epoch
        self.seq = seq
        self.tWrite = tWrite
    }
}

public struct OrbeliasHit: Equatable, Sendable {
    public let id: OrbeliasId
    public let score: Double
    public let payload: Data
}

public struct RetrieveStats: Equatable, Sendable {
    public let totalVectors: Int
    public let scanned: Int
    public let skippedVectors: Int
    public let returned: Int
    public let ringsTouched: Int
    public let payloadBytes: Int
    public let estimatedTokens: Int
    public let fanoutNodes: Int
    public let candidateReduction: Double
}

public struct RetrieveResult: Equatable, Sendable {
    public let hits: [OrbeliasHit]
    public let stats: RetrieveStats
}

public final class OrbeliasDB {
    public static let abiVersion: Int32 = 2
    private static let runtimeInit: Void = {
        orbelias_init()
    }()

    private var handle: UnsafeMutableRawPointer?

    private init(handle: UnsafeMutableRawPointer) {
        self.handle = handle
    }

    deinit {
        close()
    }

    public static func open(nodes: Int32 = 8) throws -> OrbeliasDB {
        try ensureRuntime()
        guard let handle = orbelias_open(nodes) else {
            throw lastError()
        }
        return OrbeliasDB(handle: handle)
    }

    public static func openDir(nodes: Int32 = 8, dir: String) throws -> OrbeliasDB {
        try ensureRuntime()
        return try dir.withCString { cDir in
            guard let handle = orbelias_open_dir(nodes, cDir) else {
                throw lastError()
            }
            return OrbeliasDB(handle: handle)
        }
    }

    public static func connectAuth(
        peers: String,
        username: String = "",
        password: String = "",
        authToken: String = "",
        secretKey: String = "",
        galaxy: String = ""
    ) throws -> OrbeliasDB {
        try ensureRuntime()
        return try peers.withCString { cPeers in
            try username.withCString { cUsername in
                try password.withCString { cPassword in
                    try authToken.withCString { cAuthToken in
                        try secretKey.withCString { cSecretKey in
                            try galaxy.withCString { cGalaxy in
                                guard let handle = orbelias_connect_auth(
                                    cPeers,
                                    cUsername,
                                    cPassword,
                                    cAuthToken,
                                    cSecretKey,
                                    cGalaxy
                                ) else {
                                    throw lastError()
                                }
                                return OrbeliasDB(handle: handle)
                            }
                        }
                    }
                }
            }
        }
    }

    public static func connect(peers: String) throws -> OrbeliasDB {
        try connectAuth(peers: peers)
    }

    public func close() {
        if let handle {
            orbelias_close(handle)
            self.handle = nil
        }
    }

    public func configureRing(_ ring: String, period: Double) throws {
        try withHandle { handle in
            try ring.withCString { cRing in
                try check(orbelias_ring_configure(handle, cRing, period))
            }
        }
    }

    public func setGalaxyDescription(_ description: String) throws {
        try withHandle { handle in
            try description.withCString { cDescription in
                try check(orbelias_set_galaxy_description(handle, cDescription))
            }
        }
    }

    public func setRingDescription(_ ring: String, description: String) throws {
        try withHandle { handle in
            try ring.withCString { cRing in
                try description.withCString { cDescription in
                    try check(orbelias_set_ring_description(handle, cRing, cDescription))
                }
            }
        }
    }

    public func put(_ ring: String, payload: Data) throws -> OrbeliasId {
        try withHandle { handle in
            try ring.withCString { cRing in
                var id = orbelias_id()
                let rc = payload.withUnsafeBytes { bytes in
                    orbelias_put(handle, cRing, bytes.baseAddress, payload.count, &id)
                }
                try check(rc)
                return OrbeliasId(c: id)
            }
        }
    }

    public func put(_ ring: String, payload: String) throws -> OrbeliasId {
        try put(ring, payload: Data(payload.utf8))
    }

    public func putVec(_ ring: String, payload: Data, vector: [Float]) throws -> OrbeliasId {
        try withHandle { handle in
            try ring.withCString { cRing in
                var id = orbelias_id()
                let rc = payload.withUnsafeBytes { payloadBytes in
                    vector.withUnsafeBufferPointer { vectorBytes in
                        orbelias_put_vec(
                            handle,
                            cRing,
                            payloadBytes.baseAddress,
                            payload.count,
                            vectorBytes.baseAddress,
                            vector.count,
                            &id
                        )
                    }
                }
                try check(rc)
                return OrbeliasId(c: id)
            }
        }
    }

    public func putVec(_ ring: String, payload: String, vector: [Float]) throws -> OrbeliasId {
        try putVec(ring, payload: Data(payload.utf8), vector: vector)
    }

    public func get(_ id: OrbeliasId) throws -> Data? {
        try withHandle { handle in
            var len = 0
            guard let ptr = orbelias_get(handle, id.cValue, &len) else {
                let err = Self.lastError()
                if case let .cAbi(message) = err, message.contains("not found") {
                    return nil
                }
                throw err
            }
            defer { orbelias_free(ptr) }
            return Data(bytes: ptr, count: len)
        }
    }

    public func getString(_ id: OrbeliasId) throws -> String? {
        guard let data = try get(id) else {
            return nil
        }
        guard let value = String(data: data, encoding: .utf8) else {
            throw OrbeliasDBError.invalidUtf8
        }
        return value
    }

    public func batchGet(_ ids: [OrbeliasId]) throws -> [Data] {
        try withHandle { handle in
            var rawIds = ids.map(\.cValue)
            let count = rawIds.count
            let result = rawIds.withUnsafeMutableBufferPointer { buffer in
                orbelias_batch_get(handle, buffer.baseAddress, count)
            }
            guard let result else {
                throw Self.lastError()
            }
            defer { orbelias_batch_get_free(result) }
            guard let values = result.pointee.values else {
                return []
            }
            return (0..<Int(result.pointee.len)).map { index in
                let value = values.advanced(by: index).pointee
                guard let data = value.data, value.len > 0 else {
                    return Data()
                }
                return Data(bytes: data, count: value.len)
            }
        }
    }

    public func query(_ id: OrbeliasId, selection: String) throws -> Data {
        try withHandle { handle in
            try selection.withCString { cSelection in
                var len = 0
                guard let ptr = orbelias_query(handle, id.cValue, cSelection, &len) else {
                    throw Self.lastError()
                }
                defer { orbelias_free(ptr) }
                return Data(bytes: ptr, count: len)
            }
        }
    }

    public func queryString(_ id: OrbeliasId, selection: String) throws -> String {
        let data = try query(id, selection: selection)
        guard let value = String(data: data, encoding: .utf8) else {
            throw OrbeliasDBError.invalidUtf8
        }
        return value
    }

    public func retrieve(
        vector: [Float],
        ring: String = "",
        budget: Int32 = 8,
        topRings: Int32 = 0,
        focus: Int32 = 0
    ) throws -> RetrieveResult {
        try withHandle { handle in
            try ring.withCString { cRing in
                let result = vector.withUnsafeBufferPointer { vectorBytes in
                    orbelias_retrieve(
                        handle,
                        vectorBytes.baseAddress,
                        vector.count,
                        cRing,
                        budget,
                        topRings,
                        focus
                    )
                }
                guard let result else {
                    throw Self.lastError()
                }
                defer { orbelias_retrieve_free(result) }
                let raw = result.pointee
                let hits: [OrbeliasHit]
                if let rawHits = raw.hits {
                    hits = (0..<Int(raw.len)).map { index in
                        let hit = rawHits.advanced(by: index).pointee
                        let payload: Data
                        if let ptr = hit.payload, hit.payload_len > 0 {
                            payload = Data(bytes: ptr, count: hit.payload_len)
                        } else {
                            payload = Data()
                        }
                        return OrbeliasHit(id: OrbeliasId(c: hit.id), score: hit.score, payload: payload)
                    }
                } else {
                    hits = []
                }
                return RetrieveResult(
                    hits: hits,
                    stats: RetrieveStats(
                        totalVectors: Int(raw.total_vectors),
                        scanned: Int(raw.scanned),
                        skippedVectors: Int(raw.skipped_vectors),
                        returned: Int(raw.returned),
                        ringsTouched: Int(raw.rings_touched),
                        payloadBytes: Int(raw.payload_bytes),
                        estimatedTokens: Int(raw.estimated_tokens),
                        fanoutNodes: Int(raw.fanout_nodes),
                        candidateReduction: raw.candidate_reduction
                    )
                )
            }
        }
    }

    public func atlas(queryVector: [Float] = [], maxCentroidDims: Int32 = 8) throws -> String {
        try withHandle { handle in
            var len = 0
            let ptr = queryVector.withUnsafeBufferPointer { buffer in
                orbelias_atlas(handle, buffer.baseAddress, queryVector.count, maxCentroidDims, &len)
            }
            guard let ptr else {
                throw Self.lastError()
            }
            defer { orbelias_free(ptr) }
            guard let value = String(data: Data(bytes: ptr, count: len), encoding: .utf8) else {
                throw OrbeliasDBError.invalidUtf8
            }
            return value
        }
    }

    public func locate(_ id: OrbeliasId, at: Double = -1.0) throws -> Int {
        try withHandle { handle in
            let node = orbelias_locate(handle, id.cValue, at)
            if node < 0 {
                throw Self.lastError()
            }
            return Int(node)
        }
    }

    public func nextVisit(_ id: OrbeliasId, node: Int32) throws -> Double {
        try withHandle { handle in
            let time = orbelias_next_visit(handle, id.cValue, node)
            if time < 0 {
                throw Self.lastError()
            }
            return time
        }
    }

    public func nextJoin(_ a: OrbeliasId, _ b: OrbeliasId) throws -> Double? {
        try withHandle { handle in
            let time = orbelias_next_join(handle, a.cValue, b.cValue)
            return time < 0 ? nil : time
        }
    }

    private static func ensureRuntime() throws {
        _ = runtimeInit
        let actual = orbelias_abi_version()
        if actual != abiVersion {
            throw OrbeliasDBError.abiVersionMismatch(expected: abiVersion, actual: actual)
        }
    }

    private static func lastError() -> OrbeliasDBError {
        guard let ptr = orbelias_last_error() else {
            return .cAbi("OrbeliasDB C ABI error")
        }
        let message = String(cString: ptr)
        return .cAbi(message.isEmpty ? "OrbeliasDB C ABI error" : message)
    }

    private func withHandle<T>(_ body: (UnsafeMutableRawPointer) throws -> T) throws -> T {
        guard let handle else {
            throw OrbeliasDBError.closed
        }
        return try body(handle)
    }

    private func check(_ code: Int32) throws {
        if code != ORBELIAS_OK {
            throw Self.lastError()
        }
    }
}

private extension OrbeliasId {
    init(c: orbelias_id) {
        self.init(parent: c.parent, epoch: c.epoch, seq: c.seq, tWrite: c.t_write)
    }

    var cValue: orbelias_id {
        orbelias_id(parent: parent, epoch: epoch, seq: seq, t_write: tWrite)
    }
}

File: drivers/php/src/OrbeliasDB.php
<?php
declare(strict_types=1);

namespace OrbeliasDB;

use FFI;
use FFI\CData;
use RuntimeException;

final class OrbeliasId
{
    public function __construct(
        public readonly int|string $parent,
        public readonly int $epoch,
        public readonly int $seq,
        public readonly float $tWrite,
    ) {
    }
}

final class OrbeliasHit
{
    public function __construct(
        public readonly OrbeliasId $id,
        public readonly float $score,
        public readonly string $payload,
    ) {
    }
}

final class RetrieveResult
{
    /** @param list<OrbeliasHit> $hits */
    public function __construct(
        public readonly array $hits,
        public readonly array $stats,
    ) {
    }
}

final class OrbeliasDB
{
    private const ABI_VERSION = 2;

    private static ?FFI $ffi = null;
    private ?CData $handle;

    private function __construct(CData $handle)
    {
        $this->handle = $handle;
    }

    public function __destruct()
    {
        $this->close();
    }

    public static function open(int $nodes = 8, string $lib = __DIR__ . '/../../../lib/liborbeliasdb.so'): self
    {
        $ffi = self::ffi($lib);
        $handle = $ffi->orbelias_open($nodes);
        if ($handle === null) {
            throw self::lastError();
        }
        return new self($handle);
    }

    public static function openDir(int $nodes, string $dir, string $lib = __DIR__ . '/../../../lib/liborbeliasdb.so'): self
    {
        $ffi = self::ffi($lib);
        $handle = $ffi->orbelias_open_dir($nodes, $dir);
        if ($handle === null) {
            throw self::lastError();
        }
        return new self($handle);
    }

    public static function connectAuth(
        string $peers,
        string $username = '',
        string $password = '',
        string $authToken = '',
        string $secretKey = '',
        string $galaxy = '',
        string $lib = __DIR__ . '/../../../lib/liborbeliasdb.so',
    ): self {
        $ffi = self::ffi($lib);
        $handle = $ffi->orbelias_connect_auth($peers, $username, $password, $authToken, $secretKey, $galaxy);
        if ($handle === null) {
            throw self::lastError();
        }
        return new self($handle);
    }

    public static function connect(string $peers, string $lib = __DIR__ . '/../../../lib/liborbeliasdb.so'): self
    {
        return self::connectAuth($peers, lib: $lib);
    }

    public function close(): void
    {
        if ($this->handle !== null) {
            self::ffi()->orbelias_close($this->handle);
            $this->handle = null;
        }
    }

    public function configureRing(string $ring, float $period): void
    {
        $this->check(self::ffi()->orbelias_ring_configure($this->requireHandle(), $ring, $period));
    }

    public function setGalaxyDescription(string $description): void
    {
        $this->check(self::ffi()->orbelias_set_galaxy_description($this->requireHandle(), $description));
    }

    public function setRingDescription(string $ring, string $description): void
    {
        $this->check(self::ffi()->orbelias_set_ring_description($this->requireHandle(), $ring, $description));
    }

    public function put(string $ring, string $payload): OrbeliasId
    {
        $ffi = self::ffi();
        $id = $ffi->new('orbelias_id');
        $this->check($ffi->orbelias_put($this->requireHandle(), $ring, $payload, strlen($payload), FFI::addr($id)));
        return self::idFromC($id);
    }

    /** @param list<float|int> $vector */
    public function putVec(string $ring, string $payload, array $vector): OrbeliasId
    {
        $ffi = self::ffi();
        $id = $ffi->new('orbelias_id');
        $vec = self::floatArray($vector);
        $vecPtr = count($vector) === 0 ? null : FFI::addr($vec[0]);
        $this->check($ffi->orbelias_put_vec(
            $this->requireHandle(),
            $ring,
            $payload,
            strlen($payload),
            $vecPtr,
            count($vector),
            FFI::addr($id),
        ));
        return self::idFromC($id);
    }

    public function get(OrbeliasId $id): ?string
    {
        $ffi = self::ffi();
        $len = $ffi->new('size_t[1]');
        $ptr = $ffi->orbelias_get($this->requireHandle(), self::idToC($id), FFI::addr($len[0]));
        if ($ptr === null) {
            $message = self::lastError()->getMessage();
            if (str_contains($message, 'not found')) {
                return null;
            }
            throw new RuntimeException($message);
        }
        try {
            return FFI::string($ptr, (int)$len[0]);
        } finally {
            $ffi->orbelias_free($ptr);
        }
    }

    /** @param list<OrbeliasId> $ids @return list<string> */
    public function batchGet(array $ids): array
    {
        $ffi = self::ffi();
        $arr = $ffi->new('orbelias_id[' . max(1, count($ids)) . ']');
        foreach ($ids as $i => $id) {
            $arr[$i] = self::idToC($id);
        }
        $res = $ffi->orbelias_batch_get($this->requireHandle(), count($ids) === 0 ? null : FFI::addr($arr[0]), count($ids));
        if ($res === null) {
            throw self::lastError();
        }
        try {
            $out = [];
            for ($i = 0; $i < (int)$res->len; $i++) {
                $value = $res->values[$i];
                $out[] = $value->data === null ? '' : FFI::string($value->data, (int)$value->len);
            }
            return $out;
        } finally {
            $ffi->orbelias_batch_get_free($res);
        }
    }

    public function query(OrbeliasId $id, string $selection): string
    {
        $ffi = self::ffi();
        $len = $ffi->new('size_t[1]');
        $ptr = $ffi->orbelias_query($this->requireHandle(), self::idToC($id), $selection, FFI::addr($len[0]));
        if ($ptr === null) {
            throw self::lastError();
        }
        try {
            return FFI::string($ptr, (int)$len[0]);
        } finally {
            $ffi->orbelias_free($ptr);
        }
    }

    /** @param list<float|int> $vector */
    public function retrieve(array $vector, string $ring = '', int $budget = 8, int $topRings = 0, int $focus = 0): RetrieveResult
    {
        $ffi = self::ffi();
        $vec = self::floatArray($vector);
        $res = $ffi->orbelias_retrieve(
            $this->requireHandle(),
            count($vector) === 0 ? null : FFI::addr($vec[0]),
            count($vector),
            $ring,
            $budget,
            $topRings,
            $focus,
        );
        if ($res === null) {
            throw self::lastError();
        }
        try {
            $hits = [];
            for ($i = 0; $i < (int)$res->len; $i++) {
                $hit = $res->hits[$i];
                $hits[] = new OrbeliasHit(
                    self::idFromC($hit->id),
                    (float)$hit->score,
                    $hit->payload === null ? '' : FFI::string($hit->payload, (int)$hit->payload_len),
                );
            }
            return new RetrieveResult($hits, [
                'totalVectors' => (int)$res->total_vectors,
                'scanned' => (int)$res->scanned,
                'skippedVectors' => (int)$res->skipped_vectors,
                'returned' => (int)$res->returned,
                'ringsTouched' => (int)$res->rings_touched,
                'payloadBytes' => (int)$res->payload_bytes,
                'estimatedTokens' => (int)$res->estimated_tokens,
                'fanoutNodes' => (int)$res->fanout_nodes,
                'candidateReduction' => (float)$res->candidate_reduction,
            ]);
        } finally {
            $ffi->orbelias_retrieve_free($res);
        }
    }

    /** @param list<float|int> $queryVector */
    public function atlas(array $queryVector = [], int $maxCentroidDims = 8): string
    {
        $ffi = self::ffi();
        $vec = self::floatArray($queryVector);
        $len = $ffi->new('size_t[1]');
        $ptr = $ffi->orbelias_atlas(
            $this->requireHandle(),
            count($queryVector) === 0 ? null : FFI::addr($vec[0]),
            count($queryVector),
            $maxCentroidDims,
            FFI::addr($len[0]),
        );
        if ($ptr === null) {
            throw self::lastError();
        }
        try {
            return FFI::string($ptr, (int)$len[0]);
        } finally {
            $ffi->orbelias_free($ptr);
        }
    }

    public function locate(OrbeliasId $id, float $at = -1.0): int
    {
        $node = self::ffi()->orbelias_locate($this->requireHandle(), self::idToC($id), $at);
        if ($node < 0) {
            throw self::lastError();
        }
        return (int)$node;
    }

    public function nextVisit(OrbeliasId $id, int $node): float
    {
        $time = self::ffi()->orbelias_next_visit($this->requireHandle(), self::idToC($id), $node);
        if ($time < 0) {
            throw self::lastError();
        }
        return (float)$time;
    }

    public function nextJoin(OrbeliasId $a, OrbeliasId $b): ?float
    {
        $time = self::ffi()->orbelias_next_join($this->requireHandle(), self::idToC($a), self::idToC($b));
        return $time < 0 ? null : (float)$time;
    }

    private function check(int $code): void
    {
        if ($code !== 0) {
            throw self::lastError();
        }
    }

    private function requireHandle(): CData
    {
        if ($this->handle === null) {
            throw new RuntimeException('OrbeliasDB handle is closed');
        }
        return $this->handle;
    }

    /** @param list<float|int> $values */
    private static function floatArray(array $values): CData
    {
        $ffi = self::ffi();
        $arr = $ffi->new('float[' . max(1, count($values)) . ']');
        foreach ($values as $i => $value) {
            $arr[$i] = (float)$value;
        }
        return $arr;
    }

    private static function idToC(OrbeliasId $id): CData
    {
        $c = self::ffi()->new('orbelias_id');
        $c->parent = $id->parent;
        $c->epoch = $id->epoch;
        $c->seq = $id->seq;
        $c->t_write = $id->tWrite;
        return $c;
    }

    private static function idFromC(CData $id): OrbeliasId
    {
        return new OrbeliasId((string)$id->parent, (int)$id->epoch, (int)$id->seq, (float)$id->t_write);
    }

    private static function lastError(): RuntimeException
    {
        $ptr = self::ffi()->orbelias_last_error();
        $message = $ptr === null ? 'OrbeliasDB C ABI error' : FFI::string($ptr);
        return new RuntimeException($message === '' ? 'OrbeliasDB C ABI error' : $message);
    }

    private static function ffi(string $lib = __DIR__ . '/../../../lib/liborbeliasdb.so'): FFI
    {
        if (!class_exists(FFI::class)) {
            throw new RuntimeException('PHP FFI extension is not enabled');
        }
        if (self::$ffi === null) {
            self::$ffi = FFI::cdef(self::CDEF, $lib);
            self::$ffi->orbelias_init();
            if ((int)self::$ffi->orbelias_abi_version() !== self::ABI_VERSION) {
                throw new RuntimeException('OrbeliasDB ABI version mismatch');
            }
        }
        return self::$ffi;
    }

    private const CDEF = <<<'CDEF'
typedef unsigned long size_t;
typedef unsigned long uint64_t;
typedef unsigned int uint32_t;

typedef struct orbelias_id {
  uint64_t parent;
  uint32_t epoch;
  uint32_t seq;
  double   t_write;
} orbelias_id;

typedef struct orbelias_hit {
  orbelias_id id;
  double   score;
  void    *payload;
  size_t   payload_len;
} orbelias_hit;

typedef struct orbelias_retrieve_result {
  size_t     len;
  orbelias_hit *hits;
  int        total_vectors;
  int        scanned;
  int        skipped_vectors;
  int        returned;
  int        rings_touched;
  int        payload_bytes;
  int        estimated_tokens;
  int        fanout_nodes;
  double     candidate_reduction;
} orbelias_retrieve_result;

typedef struct orbelias_value {
  void  *data;
  size_t len;
} orbelias_value;

typedef struct orbelias_batch_result {
  size_t       len;
  orbelias_value *values;
} orbelias_batch_result;

int         orbelias_abi_version(void);
const char *orbelias_last_error(void);
void        orbelias_init(void);
void       *orbelias_open(int nodes);
void       *orbelias_open_dir(int nodes, const char *dir);
void       *orbelias_connect_auth(const char *peers, const char *username, const char *password, const char *auth_token, const char *secret_key, const char *galaxy);
void        orbelias_close(void *db);
int         orbelias_ring_configure(void *db, const char *ring, double period);
int         orbelias_set_galaxy_description(void *db, const char *description);
int         orbelias_set_ring_description(void *db, const char *ring, const char *description);
int         orbelias_put(void *db, const char *ring, const void *data, size_t len, orbelias_id *out_id);
int         orbelias_put_vec(void *db, const char *ring, const void *data, size_t len, const float *vec, size_t vec_len, orbelias_id *out_id);
void       *orbelias_get(void *db, orbelias_id id, size_t *out_len);
void        orbelias_free(void *p);
orbelias_batch_result *orbelias_batch_get(void *db, const orbelias_id *ids, size_t ids_len);
void        orbelias_batch_get_free(orbelias_batch_result *r);
void       *orbelias_query(void *db, orbelias_id id, const char *selection, size_t *out_len);
orbelias_retrieve_result *orbelias_retrieve(void *db, const float *vec, size_t vec_len, const char *ring, int budget, int top_rings, int focus);
void        orbelias_retrieve_free(orbelias_retrieve_result *r);
void       *orbelias_atlas(void *db, const float *query_vec, size_t query_vec_len, int max_centroid_dims, size_t *out_len);
int         orbelias_locate(void *db, orbelias_id id, double at);
double      orbelias_next_visit(void *db, orbelias_id id, int node);
double      orbelias_next_join(void *db, orbelias_id a, orbelias_id b);
CDEF;
}
